### PR TITLE
Fix cross join lose predicate in JoinAssociativityRule

### PR DESCRIPTION
**Pull request by Seaven**
Such as SQL:
```
SELECT *
FROM (
	SELECT CASE 
			WHEN t0.v3 THEN t2.v2
			WHEN s0.v3 THEN '123'
			ELSE t2.v3
		END AS x0, t1.v1
	FROM s0
		JOIN t0 ON s0.v2 = t0.v2
		JOIN t1 ON s0.v2 = t1.v2 OR s0.v3 = t1.v2
		JOIN t2 ON t1.v2 = t2.v2
	WHERE t2.v1 = 2
) temp
WHERE x0 > 1;
```

The Join plan:
```
|   15:Project                                                                                                                                                 |
|   |  <slot 7> : 7: v1                                                                                                                                        |
|   |  <slot 13> : CASE WHEN CAST(6: v3 AS BOOLEAN) THEN CAST(11: v2 AS VARCHAR) WHEN CAST(3: v3 AS BOOLEAN) THEN '123' ELSE CAST(12: v3 AS VARCHAR) END       |
|   |                                                                                                                                                          |
|   14:CROSS JOIN                                                                                                                                              |
|   |  cross join:                                                                                                                                             |
|   |  predicates: CASE WHEN CAST(6: v3 AS BOOLEAN) THEN CAST(11: v2 AS VARCHAR) WHEN CAST(3: v3 AS BOOLEAN) THEN '123' ELSE CAST(12: v3 AS VARCHAR) END > '1' |
|   |                                                                                                                                                          |
|   |----13:EXCHANGE                                                                                                                                           |
|   |                                                                                                                                                          |
|   6:Project                                                                                                                                                  |
|   |  <slot 2> : 2: v2                                                                                                                                        |
|   |  <slot 3> : 3: v3                                                                                                                                        |
|   |  <slot 6> : 6: v3                                                                                                                                        |
|   |                                                                                                                                                          |
|   5:HASH JOIN                                                                                                                                                |
|   |  join op: INNER JOIN (BROADCAST)                                                                                                                         |
|   |  hash predicates:                                                                                                                                        |
|   |  colocate: false, reason:                                                                                                                                |
|   |  equal join conjunct: 15: cast = 14: cast
```

The plan was lose predicate `s0.v2 = t1.v2 OR s0.v3 = t1.v2` on CROSS_JOIN

The right plan:
```
|   14:CROSS JOIN                                                                                                                                              |
|   |  cross join:                                                                                                                                             |
|   |  predicates: (CAST(2: v2 AS DOUBLE) = CAST(8: v2 AS DOUBLE)) OR (3: v3 = 8: v2), CASE WHEN CAST(6: v3 AS BOOLEAN) THEN CAST(11: v2 AS VARCHAR) WHEN CAST(3: v3 AS BOOLEAN) THEN '123' ELSE CAST(12: v3 AS VARCHAR) END > '1' |
```

### CODE REVIEW

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java [hunk 1]
```diff
@@ -242,4 +259,29 @@ public void testTPCDS22() throws Exception {
                 "  |----3:EXCHANGE\n" +
                 "  |       cardinality: 335"));
     }
+
+    @Test
+    public void testCrossReorder() throws Exception {
+        RuleSet mockRule = new RuleSet() {
+            @Override
+            public void addJoinTransformationRules() {
```
Comment (Youngwb) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java:267]: why mock the rule

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java [hunk 2]
```diff
@@ -126,6 +131,18 @@ private String getDumpInfoFromFile(String fileName) throws Exception {
                         getExplainString(TExplainLevel.COSTS));
     }
 
+    private Pair<QueryDumpInfo, String> getPlanFragment(String dumpJsonStr, SessionVariable sessionVariable,
```
Comment (Youngwb) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java:134]: It looks like getCostPlanFragment，you can modify getCostPlanFragment directly

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java:267]: Avoid join reorder more plan

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java:134]: Yeah, but you know, choose the plan by mock cost is really troublesome

**Review by Youngwb — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**